### ISSUE TITLE
The folder numbers during the training process are not in order

### ISSUE THREAD

**Issue by jaylongnwnr** [ultralytics/ultralytics#6620]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

As I said before, the first training will be 1, the second training will be 2, and the third training will not be 3 but 22. I have observed this situation many times, so I raised this question and asked the engineer to look at it. one time

It was originally possible to train, but after wiping it again and updating the version of YOLOV8, it became impossible to train.

Because I am currently doing tests related to federated learning, this situation only occurs when using federated learning. Simply using YOLOV8 for training does not have this problem.

But I'm using the same code, just without using the federated learning framework to wrap it.

<img width="2621" alt="截圖 2023-11-28 11 28 36" src="https://github.com/ultralytics/ultralytics/assets/32053921/7175fbf8-9340-48d8-a303-7537b92c681b">
<img width="2619" alt="截圖 2023-11-28 11 29 02" src="https://github.com/ultralytics/ultralytics/assets/32053921/6177ea73-56f3-4319-97a6-5c4f806649ae">

It's because the file name becomes 22 upside down and the folder I want to capture 3 cannot be captured, so the whole thing will crash.

@glenn-jocher

**Comment by glenn-jocher** [comment]
@jaylongnwnr hi there! Thanks for reaching out with your question regarding the directory numbering during the federated learning process using YOLOv8.

It's interesting to hear that you're encountering irregularities in the folder numbering sequence when training your models in a federated learning setup. The model training directories are typically created in a sequential order for each new training session. If you are experiencing out-of-sequence numbering such as jumping from 2 to 22, this could suggest that there might be some state retention or unexpected behavior triggered by the federated learning wrapping, which may be influencing how the directories are being generated and named.

Without code or access to your specific setup, I can only advise you to closely inspect the part of your code that interfaces with the YOLOv8 training process. Specifically, look for any operations that could potentially interfere with the directory naming convention in YOLOv8. Since you mentioned this does not occur when not using the federated learning framework, the issue likely lies there.

In addition, ensure that you're using the latest version of YOLOv8 and consider reverting to a previous stable state where the directory numbering was consistent, then carefully reintroduce changes relevant for federated learning.

If the problem persists, consider raising an issue with the federated learning framework you're using, as they may have insights on how their system manipulates or interacts with file structures during its operations.

For more detailed guidance on the training process itself or managing the training output, you can refer to the Training section of the Ultralytics Docs.

Best of luck with your federated learning tests, and I'm sure with a systematic approach, you'll be able to get to the bottom of this! 😊🔍

**Comment by glenn-jocher** [comment]
@jaylongnwnr Hi again!

Ah, if you started experiencing this issue only after updating YOLOv8, it might indeed be related to changes introduced in the newer version. While downgrading to a previous version where you didn't experience this problem could be a temporary fix, it would be good to understand the root cause associated with the update.

Nevertheless, reverting to an earlier version is a reasonable step to quickly get your federated learning tests back on track, as it could eliminate the immediate problem. Just keep in mind that this is a temporary solution and you may miss out on improvements and bug fixes introduced in the latest versions.

To proceed with a more long-term solution, you could do the following:
1. **Isolate the problem**: Ensure that the issue is reproducible on the updated version and absent on the previous version under the same conditions.
2. **Check the release notes**: Review the release notes for the version you updated to; sometimes, they document changes that could affect your workflow.
3. **Raise an issue**: If after isolating the problem, you've confirmed that the update introduced the issue, consider raising an issue on the Ultralytics YOLOv8 GitHub repository with detailed information. This way, if it's a bug, the team can work on a fix for future releases.

In any scenario, your feedback can be valuable for the Ultralytics team to address potential bugs or document changes that might affect your use case. If you need further assistance, the Documentation might provide additional insights on managing versions and troubleshooting.

Keep up the great work with federated learning, and thanks for contributing to the continuous improvement of YOLOv8! 😊🛠️

**Comment by jaylongnwnr** [comment]
@glenn-jocher  hi,
So how do I downgrade the yolov8 version? I didn’t find any relevant documentation.

**Comment by jaylongnwnr** [comment]
@glenn-jocher hi again,

After I lowered the version, the problem of train3 has been solved.

But there will be a problem that it will kill itself halfway through the training process.

I would like to ask what is the problem?

**Comment by glenn-jocher** [comment]
@jaylongnwnr Hi!

It's good to hear that downgrading the version resolved the directory numbering issue. Regarding the self-termination problem during training, it can occur due to a variety of reasons. Here are some common ones to consider:

1. **Hardware limitations**: Training might stop if there's insufficient memory (RAM or GPU VRAM) or if the system is overheating. Ensure your hardware meets the requirements for the training task at hand and is properly cooled.

2. **Software conflicts**: Ensure that there are no conflicting processes on your machine that might interrupt the training. Also, verify that your CUDA and cuDNN versions are compatible with the version of YOLOv8 you are using.

3. **Resource limits**: If you're training on a shared system or in an environment with resource quotas (like cloud platforms), make sure you aren't hitting any resource usage limits.

4. **Bug in the code**: There might be a bug introduced in the version you downgraded to which is causing the process to terminate. Check the YOLOv8 GitHub Issues page to see if anyone has reported something similar for that version.

5. **Improper configuration**: Misconfigured training settings can cause the training to stop. Double-check your configuration files to make sure everything is set up correctly.

6. **Logs and error messages**: Look at the error logs produced by the training process. They often provide clues to what caused the termination.

7. **System stability**: Ensure your operating system is stable and up-to-date. Sometimes, system updates or driver issues can cause unexpected behavior during intensive computational tasks.

To help remedy the situation:
- Monitor system resources (CPU, GPU, memory) usage during training to identify potential bottlenecks.
- Make sure other non-essential processes are not running on the system that might interfere with training.
- Check terminal output or log files for error messages that can offer more insight.

If you're still unable to diagnose the problem, you can reach out to the Ultralytics team with detailed logs and a description of your setup, and they might be able to provide additional support.

Keep experimenting and don't hesitate to reach out if you encounter more hurdles. Your perseverance is essential in these complex tasks! 😊👍

**Comment by jaylongnwnr** [comment]
@glenn-jocher hi agian,

I have two types of data that need to be added to training
For example, if my batch is set to 8, I want to draw 4 images for each folder .
May I ask what method can be used to accomplish this?

**Comment by glenn-jocher** [comment]
@jaylongnwnr Hello!

If you're looking to train your YOLOv8 model with data coming from two different folders and want to have an equal number of images selected from each folder for every batch during training, you would need to implement a custom data loading mechanism.

To accomplish this, you could:
1. **Modify the Dataset Loader**: Customize the dataset loader so that it alternates between the two datasets while fetching data for a new batch. This custom loader would need to ensure that for each batch, it draws an equal number of images from both folders.

2. **Merge Datasets**: Another approach could be to physically merge your datasets into a single folder with a consistent annotation format. Be sure to mix the files from both sources so that each batch naturally contains images from both datasets.

3. **Custom Batching Function**: Write a custom batching function that interfaces with the model's data loader. This function should extract the specified number of images from each dataset for every batch.

Keep in mind that handling datasets and batching could require in-depth modifications to the data pipeline, so it may not be trivial depending on your level of experience with YOLOv8's codebase. Always start by reviewing the documentation which might offer insights into handling multiple datasets.

Remember that the changes should preserve the randomness of data selection within each folder to maintain the generalization ability of the model.

Implementing such a feature may require a good understanding of PyTorch's `DataLoader` and `Sampler` classes, or you might use custom code to manage the batch assembly process.

I hope this gives you an idea of the direction to head towards. Good luck with your training! 😊🚀

**Comment by jaylongnwnr** [comment]
@glenn-jocher 
Hi, I'm sorry to trouble you again.
I'm still a beginner. Is there any relevant literature you can provide me with such a similar approach, such as sample programs?
Because I think the sampling effect of putting the data together is not good, I am currently using this left method, but the test effect of a certain data set is not good.

**Comment by jaylongnwnr** [comment]
@glenn-jocher hi,

Plotting labels to runs/detect/train3/labels.jpg... 
Killed
student1@SCL-TUF:~/Documents/FL_yolo_test/tets$ /usr/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 115 leaked semaphore objects to clean up at shutdown
warnings.warn('resource_tracker: There appear to be %d '

I have this error now. Is this YOLOV8 related?

**Comment by glenn-jocher** [comment]
@jaylongnwnr Hello!

The message you're seeing indicates that your training process was terminated ("Killed"), and following that, there is a warning about leaked semaphore objects, which is likely a side-effect of the abrupt termination rather than the root cause.

The main issue to focus on is why the process was killed. This is typically caused by the system running out of resources, such as memory. Here are some common reasons for the termination:

- **Out of Memory**: The training may have consumed all available system or GPU memory, which can trigger the operating system to kill the process to maintain system stability.
- **System Resource Limits**: You might hit a resource limit if you're using a system that imposes such constraints, like with certain cloud services or managed environments.

The leaked semaphore objects warning refers to resources that weren't properly released because of the termination. This is secondary to the fundamental issue of the process being killed.

To troubleshoot this, consider the following:

1. **Monitor Resources**: Use tools like `htop`, `nvidia-smi`, or equivalent to monitor system and GPU memory usage during training.
2. **Reduce Batch Size**: Decreasing the batch size can lower memory usage.
3. **Close Other Applications**: Make sure no unnecessary applications are running that could consume memory.
4. **Incremental Testing**: Try running simpler, smaller scale training jobs to ensure stability before scaling up.

If you determine that memory is the issue and you are unable to reduce usage, you may need to consider upgrading your hardware with more memory or using a system with greater capacity to handle your training demands.

Remember that while it's important to carefully manage resources, especially with deep learning models that can be resource-intensive, don't be discouraged by these challenges—they're a normal part of the learning curve in the AI field. Good luck! 😊🔄🔍

**Comment by MUHAMMAD-ABDUL-RAFAY** [comment]
@jaylongnwnr hi there!

did you able to solve the problem of yolov8 with federated learning. Actually I am working on a project where i have to integrate federated learning with yolov8. If you have no problem can you please share the code so that I can have a head start in my project
